module/zstd/zstd.c

```diff
@@ -0,0 +1,647 @@
+static int
+zstd_enum_to_cookie(enum zio_zstd_levels level, int16_t *cookie)
+{
+	for (int i = 0; i < ARRAY_SIZE(zstd_levels); i++) {
+		if (zstd_levels[i].level == level) {
+			*cookie = zstd_levels[i].cookie;
+			return (zstd_levels[i].cookie);
+		}
+	}
+
+	/* Invalid/unknown ZSTD level - this should never happen. */
+	return (1);
+}
+
+/* Compress block using zstd */
+size_t
+zstd_compress(void *s_start, void *d_start, size_t s_len, size_t d_len,
+    int level)
+{
+	size_t c_len;
+	uint32_t bufsize;
+	int16_t levelcookie;
+	char *dest = d_start;
+	ZSTD_CCtx *cctx;
+
+	/* Skip compression if the specified level is invalid */
+	if (zstd_enum_to_cookie(level, &levelcookie)) {
+		return (s_len);
+	}
+
+	ASSERT3U(d_len, >=, sizeof (bufsize));
+	ASSERT3U(d_len, <=, s_len);
+	ASSERT3U(levelcookie, !=, 0);
+
+	cctx = ZSTD_createCCtx_advanced(zstd_malloc);
+
+	/*
+	 * Out of kernel memory, gently fall through - this will disable
+	 * compression in zio_compress_data
+	 */
+	if (!cctx) {
+		return (s_len);
+	}
+
+	/* Set the compression level */
+	ZSTD_CCtx_setParameter(cctx, ZSTD_c_compressionLevel, levelcookie);
+
+	/* Use the "magicless" zstd header which saves us 4 header bytes */
+	ZSTD_CCtx_setParameter(cctx, ZSTD_c_format, ZSTD_f_zstd1_magicless);
+
+	/*
+	 * Disable redundant checksum calculation and content size storage since
+	 * this is already done by ZFS itself.
+	 */
+	ZSTD_CCtx_setParameter(cctx, ZSTD_c_checksumFlag, 0);
+	ZSTD_CCtx_setParameter(cctx, ZSTD_c_contentSizeFlag, 0);
+
+	c_len = ZSTD_compress2(cctx,
+	    &dest[sizeof (bufsize) + sizeof (levelcookie)],
+	    d_len - sizeof (bufsize) - sizeof (levelcookie),
+	    s_start, s_len);
+
+	ZSTD_freeCCtx(cctx);
+
+	/* Error in the compression routine, disable compression. */
+	if (ZSTD_isError(c_len)) {
+		return (s_len);
+	}
+
+	/*
+	 * Encode the compressed buffer size at the start. We'll need this in
+	 * decompression to counter the effects of padding which might be added
+	 * to the compressed buffer and which, if unhandled, would confuse the
+	 * hell out of our decompression function.
+	 */
+	bufsize = c_len;
+	*(uint32_t *)dest = BE_32(bufsize);
+
+	/*
+	 * Encode the compression level as well. We may need to know the
+	 * original compression level if compressed_arc is disabled, to match
+	 * the compression settings to write this block to the L2ARC.
+	 *
+	 * Encode the actual level, so if the enum changes in the future, we
+	 * will be compatible.
+	 *
+	 * The upper 16 bits store the ZSTD version to be able to provide
+	 * future compatibility, since new versions might enhance the
+	 * compression algorithm in a way, where the compressed data will
+	 * change.
+	 *
+	 * As soon as such incompatibility occurs, handling code needs to be
+	 * added, differentiating between the versions.
+	 */
+	ASSERT3U(ZSTD_VERSION_NUMBER, <=, 0xFFFF);
+	*(uint32_t *)(&dest[sizeof (bufsize)]) =
+	    BE_32(level | (ZSTD_VERSION_NUMBER << 16));
```

Comment:
since level is just 8 bit instead of levelcookie with is signed 16 bit. you can use 24 bits for zstd_version_number. so << 8 is enough and leaves space for any future version

Reply:
That does seem to make sense here. It is not likely to ever make sense to ask the user to pick between 250+ levels, and 24 bits leaves us plenty of room for future versions

Reply:
done, max. version 1677.72.15